syntax = "proto3";

package iroha;

import "block.proto";
import "google/protobuf/empty.proto";

service Storage {
  rpc AddBlock (BlockMessage) returns (google.protobuf.Empty);
  rpc GetBlock (BlockIdMessage) returns (BlockMessage);
  rpc GetPeers (google.protobuf.Empty) returns (PeersResponse);
  rpc EraseBlock (BlockIdMessage) returns (google.protobuf.Empty);
  rpc GetBlocks (RangeGetRequest) returns (stream BlockMessage);
}

message BlockMessage {
  Block block = 1;
}

message BlockIdMessage {
  oneof id {
    uint32 height = 1;
    bytes block_hash = 2;
    bytes transaction_hash = 3;
  }
}

message RangeGetRequest {
  uint32 begin = 1;
  uint32 end = 2;
}

message PeersResponse {
  repeated string address = 1;
}
